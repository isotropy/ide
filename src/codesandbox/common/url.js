// @flow

export const getSandboxOptions = (url: string) => {
  const result: Object = {
    editorSize: 50,
  };
  const moduleMatch = url.match(/(\?|&)(module)=([^&]+)/);
  if (moduleMatch) {
    result.currentModule = moduleMatch[3];
  }

  const initialPathMatch = url.match(/(\?|&)(initialpath)=([^&]+)/);
  if (initialPathMatch) {
    result.initialPath = decodeURIComponent(initialPathMatch[3]);
  }

  const fontSizeMatch = url.match(/(\?|&)(fontsize)=([^&]+)/);
  if (fontSizeMatch) {
    result.fontSize = +fontSizeMatch[3];
  }

  const editorSizeMatch = url.match(/(\?|&)(editorsize)=([^&]+)/);
  if (editorSizeMatch) {
    result.editorSize = +editorSizeMatch[3];
  }

  result.isPreviewScreen = url.includes('view=preview');
  result.isEditorScreen = url.includes('view=editor');

  // If there is no view specified and the width of the window is <800 we want
  // to default to preview
  if (!result.isPreviewScreen && !result.isEditorScreen) {
    const windowWidth =
      window.innerWidth || document.documentElement.clientWidth;
    result.isPreviewScreen = windowWidth < 800;
  }

  result.hideNavigation = url.includes('hidenavigation=1');
  result.isInProjectView = !url.includes('moduleview=1');
  result.autoResize = url.includes('autoresize=1');
  result.useCodeMirror = url.includes('codemirror=1');
  result.enableEslint = url.includes('eslint=1');

  return result;
};
