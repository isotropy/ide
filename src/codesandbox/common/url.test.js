import { getSandboxOptions } from './url';

function testSandboxOptions(url: string) {
  expect(getSandboxOptions(url)).toMatchSnapshot();
}

describe('url parameters', () => {
  it('keeps everything false on normal urls', () => {
    testSandboxOptions('https://edit.looptype.com/s/new');
  });

  it('sets current module if there is one', () => {
    testSandboxOptions('https://edit.looptype.com/s/new?module=test');
  });

  it('sets preview view', () => {
    testSandboxOptions('https://edit.looptype.com/s/new?view=preview');
  });

  it('sets editor view', () => {
    testSandboxOptions('https://edit.looptype.com/s/new?view=editor');
  });

  it("doesn't set unknown fields", () => {
    testSandboxOptions('https://edit.looptype.com/s/new?view=both');
  });

  it('can hide navigation', () => {
    testSandboxOptions('https://edit.looptype.com/s/new?hidenavigation=1');
  });

  it('can autoresize', () => {
    testSandboxOptions('https://edit.looptype.com/s/new?autoresize=1');
  });

  it('can handle multiple options', () => {
    testSandboxOptions(
      'https://edit.looptype.com/s/new?autoresize=1&view=editor&module=test&hidenavigation=1'
    );
  });
});
