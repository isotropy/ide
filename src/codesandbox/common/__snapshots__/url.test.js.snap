// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`url parameters can autoresize 1`] = `
Object {
  "autoResize": true,
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": false,
  "isEditorScreen": false,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters can handle multiple options 1`] = `
Object {
  "autoResize": true,
  "currentModule": "test",
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": true,
  "isEditorScreen": true,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters can hide navigation 1`] = `
Object {
  "autoResize": false,
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": true,
  "isEditorScreen": false,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters doesn't set unknown fields 1`] = `
Object {
  "autoResize": false,
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": false,
  "isEditorScreen": false,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters keeps everything false on normal urls 1`] = `
Object {
  "autoResize": false,
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": false,
  "isEditorScreen": false,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters sets current module if there is one 1`] = `
Object {
  "autoResize": false,
  "currentModule": "test",
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": false,
  "isEditorScreen": false,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters sets editor view 1`] = `
Object {
  "autoResize": false,
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": false,
  "isEditorScreen": true,
  "isInProjectView": true,
  "isPreviewScreen": false,
  "useCodeMirror": false,
}
`;

exports[`url parameters sets preview view 1`] = `
Object {
  "autoResize": false,
  "editorSize": 50,
  "enableEslint": false,
  "hideNavigation": false,
  "isEditorScreen": false,
  "isInProjectView": true,
  "isPreviewScreen": true,
  "useCodeMirror": false,
}
`;
